### ISSUE TITLE
Get samples of arbitrary variable in @model, via return ?

### ISSUE THREAD

**Issue by emilemathieu** [TuringLang/Turing.jl#346]
Hi all !

I thought that the `return` statement within a `@model` was similar to `predict` in Anglican (such as `predict :theta-pt07 (> theta 0.7)`), i.e. to set up structured return values (which will be sampled).
Yet, it seems that putting a `return` statement does not change anything and that the `varInfo` object will anyway (only) keep track of all sampled random variables.

Is that intended ? What if we would like to sample according to the posterior distribution of the boolean random variable ` theta > 0.7` defined as below ?

```julia
@model beta_flip(y) = begin
  theta ~ Beta(5,3)
  y ~ Bernoulli(theta)
  return theta > 0.7
end
```

Best,
Emile

**Comment by xukai92** [comment]
Hi Emile,

Yes this issue is known and not resolved yet, originally mentioned in #132. We were trying to fix that in in #322 (in short I introduced a new filed `pred` to `VarInfo` to achieve that) with the basic function implemented but decided to disable that because of two unsolved bugs/problems:

- PG: maintaining predictions for each coroutine in `VarInfo` gives some (indeterministic) bugs
  - I don't fully understand the bug though.
- HMC: dealing with numerical errors need to be careful
  - This requires to save/restore `VarInfo` more systematically.

In short I think the HMC one is easy to resolve but PG one needs more time.

I will fix the return problem in the future but it has lower priority than stable NUTS and 0.6 compatibility. 

If you are interested in fixing the bugs you can experiment it by enabling the explicit return (I disabled it in https://github.com/yebai/Turing.jl/pull/322/commits/52958ab44558fa28f73e73b3e9a1432bad66947e).

Another problem related to your example is that we are currently only intended to support user to return **a symbol** rather than an expression, considering that we need to store that in `VarInfo`. However this is believed to be trivial because users can always write another assignment statement before the return and return the symbol.

Kai

**Comment by yebai** [comment]
> Is that intended ? What if we would like to sample according to the posterior distribution of the boolean random variable theta > 0.7 defined as below ?

Turing aims at maximal consistency with standard Julia syntax. So the `return` statement in `@model` acts the same as a `return` statement in a function definition. For example

 ```julia
using Turing

@model gdemo(x, y) = begin
         s ~ InverseGamma(2,3)
         m ~ Normal(0,sqrt(s))
         x ~ Normal(m, sqrt(s))
         y ~ Normal(m, sqrt(s))
         return s, m
       end
```

Running the model without passing any data (i.e. draw from the prior): 

```julia
julia> gdemo()()
┌ Warning: Data `y` not provided, treating as parameter instead.
└ @ Main ~/.julia/dev/Turing/src/core/compiler.jl:326
┌ Warning: Data `x` not provided, treating as parameter instead.
└ @ Main ~/.julia/dev/Turing/src/core/compiler.jl:326
[ Info:  Assume - `s` is a parameter
[ Info:  Assume - `m` is a parameter
[ Info:  Assume - `x` is a parameter
[ Info:  Assume - `y` is a parameter
(5.554127504124724, 0.37645723575465434)
```
